src/target/riscv/riscv-013.c

```diff
@@ -3396,311 +3396,532 @@ static int write_memory_abstract(struct target *target, target_addr_t address,
-			LOG_DEBUG("index=%d, reads=%d, next_index=%d, ignore_last=%d, j=%d",
-				index, reads, next_index, ignore_last, j);
-			if (j + 3 + ignore_last > next_index)
-				break;
+	info->cmderr = get_field(abstractcs, DM_ABSTRACTCS_CMDERR);
+	switch (info->cmderr) {
+	case CMDERR_NONE:
+		LOG_TARGET_DEBUG(target, "successful (partial?) memory read [%"
+				PRIu32 ", %" PRIu32 ")", start_index, start_index + elements_to_read);
+		elements_to_fetch_from_batch = elements_to_read;
+		break;
+	case CMDERR_BUSY:
+		LOG_TARGET_DEBUG(target, "memory read resulted in busy response");
+		if (read_memory_progbuf_inner_on_ac_busy(target, start_index,
+					elements_to_read, &elements_to_fetch_from_batch, access)
+				!= ERROR_OK)
+			return ERROR_FAIL;
+		break;
+	default:
+		LOG_TARGET_DEBUG(target, "error when reading memory, cmderr=0x%" PRIx32,
+				info->cmderr);
+		riscv013_clear_abstract_error(target);
+		return ERROR_FAIL;
+	}
 
-			status = riscv_batch_get_dmi_read_op(batch, read_count);
-			uint64_t value = riscv_batch_get_dmi_read_data(batch, read_count);
-			read_count++;
-			if (status != DMI_STATUS_SUCCESS) {
-				/* If we're here because of busy count, dmi_busy_delay will
-				 * already have been increased and busy state will have been
-				 * cleared in dmi_read(). */
-				/* In at least some implementations, we issue a read, and then
-				 * can get busy back when we try to scan out the read result,
-				 * and the actual read value is lost forever. Since this is
-				 * rare in any case, we return error here and rely on our
-				 * caller to reread the entire block. */
-				LOG_WARNING("Batch memory read encountered DMI error %d. "
-						"Falling back on slower reads.", status);
-				riscv_batch_free(batch);
-				result = ERROR_FAIL;
-				goto error;
-			}
-			if (size > 4) {
-				status = riscv_batch_get_dmi_read_op(batch, read_count);
-				if (status != DMI_STATUS_SUCCESS) {
-					LOG_WARNING("Batch memory read encountered DMI error %d. "
-							"Falling back on slower reads.", status);
-					riscv_batch_free(batch);
-					result = ERROR_FAIL;
-					goto error;
-				}
-				value <<= 32;
-				value |= riscv_batch_get_dmi_read_data(batch, read_count);
-				read_count++;
-			}
-			riscv_addr_t offset = j * size;
-			buf_set_u64(buffer + offset, 0, 8 * size, value);
-			log_memory_access64(address + j * increment, value, size, true);
-		}
+	if (read_memory_progbuf_inner_read_batch(target, batch, start_index,
+				elements_to_fetch_from_batch, elements_read, access) != ERROR_OK)
+		return ERROR_FAIL;
 
-		index = next_index;
+	return ERROR_OK;
+}
 
-		riscv_batch_free(batch);
+static int read_memory_progbuf_inner_loop_iteration(struct target *target,
+		struct memory_access_info access, uint32_t *index, uint32_t loop_count);
+
+static int read_memory_progbuf_inner_ensure_forward_progress(struct target *target,
+		struct memory_access_info access, uint32_t start_index, uint32_t elements_to_read)
+{
+	if (elements_to_read == 1) {
```

Comment:
Can you comment that this check is to prevent infinite recursion between read_memory_progbuf_inner_ensure_forward_progress() and read_memory_progbuf_inner_loop_iteration()?

Reply:
Thanks for the great suggestion! Addressed.